### ISSUE TITLE
Polyglot models download broken (gives a 404 error)

### ISSUE THREAD

**Issue by jacotec** [LibreTranslate/LibreTranslate#344]
I'm running LibreTranslate from the docker image. Basically it works well, translating from languages to German. But as soon as I switch to translate from German to a different language, I'm getting a 404 error.

Example:

Test from English to German:

![image](https://user-images.githubusercontent.com/3321655/201922471-9c4d00dc-4bf0-4e5b-98e0-6f661655706d.png)

When I now **just click on the arrow to swap languages**, I'm getting this:

![image](https://user-images.githubusercontent.com/3321655/201922569-e4912c8c-d2a3-47bd-9d14-d244f6774971.png)

The same appears when I manually select "German" on the left and any other language on the right. As soon as I start typing in the left field, the error appears.

**Comment by jspablo** [comment]
I am facing the same issue. Translation works from english to any language but not the other way around. 
```
{"error":"Cannot translate text: HTTP Error 404: Not Found"}
```
Got the following errors when building the docker image:
```
Updating language models
Found 56 models
Downloading Arabic → English (1.0) ...
Downloading Azerbaijani → English (1.5) ...
Downloading Chinese → English (1.7) ...
Downloading Czech → English (1.5) ...
Downloading Danish → English (1.3) ...
Downloading Dutch → English (1.4) ...
Downloading English → Arabic (1.0) ...
Downloading English → Azerbaijani (1.5) ...
Downloading English → Chinese (1.7) ...
Downloading English → Czech (1.5) ...
Downloading English → Danish (1.3) ...
Downloading English → Dutch (1.4) ...
Downloading English → Esperanto (1.5) ...
Downloading English → Finnish (1.5) ...
Downloading English → French (1.0) ...
Downloading English → German (1.0) ...
Downloading English → Greek (1.5) ...
Downloading English → Hebrew (1.5) ...
Downloading English → Hindi (1.1) ...
Downloading English → Hungarian (1.5) ...
Downloading English → Indonesian (1.2) ...
Downloading English → Irish (1.1) ...
Downloading English → Italian (1.0) ...
Downloading English → Japanese (1.1) ...
Downloading English → Korean (1.1) ...
Downloading English → Persian (1.5) ...
Downloading English → Polish (1.1) ...
Downloading English → Portuguese (1.0) ...
Downloading English → Russian (1.7) ...
Downloading English → Slovak (1.5) ...
Downloading English → Spanish (1.0) ...
Downloading English → Swedish (1.5) ...
Downloading English → Turkish (1.5) ...
Downloading English → Ukranian (1.4) ...
Downloading Esperanto → English (1.5) ...
Downloading Finnish → English (1.5) ...
Downloading French → English (1.0) ...
Downloading German → English (1.0) ...
Downloading Greek → English (1.5) ...
Downloading Hebrew → English (1.5) ...
Downloading Hindi → English (1.1) ...
Downloading Hungarian → English (1.5) ...
Downloading Indonesian → English (1.2) ...
Downloading Irish → English (1.1) ...
Downloading Italian → English (1.0) ...
Downloading Japanese → English (1.1) ...
Downloading Korean → English (1.1) ...
Downloading Persian → English (1.5) ...
Downloading Polish → English (1.1) ...
Downloading Portuguese → English (1.0) ...
Downloading Russian → English (1.0) ...
Downloading Slovak → English (1.5) ...
Downloading Spanish → English (1.0) ...
Downloading Swedish → English (1.5) ...
Downloading Turkish → English (1.1) ...
Downloading Ukranian → English (1.4) ...
Loaded support for 29 languages (56 models total)!
Installing transliteration models for the following languages: ar, az, zh, cs, da, nl, eo, fi, fr, de, el, he, hi, hu, id, ga, it, ja, ko, fa, pl, pt, ru, sk, es, sv, tr, uk
[polyglot_data] Error loading transliteration2.ar: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.az: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.zh: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.cs: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.da: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.nl: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.eo: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.fi: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.fr: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.de: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.el: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.he: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.hi: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.hu: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.id: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.ga: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.it: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.ja: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.ko: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.fa: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.pl: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.pt: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.ru: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.sk: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.es: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.sv: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.tr: HTTP Error 404: Not
[polyglot_data]     Found
[polyglot_data] Error loading transliteration2.uk: HTTP Error 404: Not
[polyglot_data]     Found
```

**Comment by PJ-Finlay** [comment]
We use Polyglot for transliteration, it looks like the German Polyglot model is failing to download.

**Comment by JpegDev** [comment]
Same problem for me. It was working one week ago.

**Comment by jspablo** [comment]
@PJ-Finlay is right, can't access the models using `polyglot`, check https://github.com/aboSamoor/polyglot/issues/267. Is there another way to download them?

**Comment by jacotec** [comment]
@jspablo Great find! Thanks for raising the issue at Polyglot!

**Comment by jspablo** [comment]
@jacotec If you have an older image probably you could extract the models from there

**Comment by jacotec** [comment]
As there are already (unanswered) questions about the maintenance status in the Polyglot repo and now obviously the download server is down it looks like LibreTranslate should host the needed files locally (maybe in GitHub) or use a different, maintained transliteration (whatever that is, I'm just a stupid user ;-) ) thing?!? @PJ-Finlay

**Comment by PJ-Finlay** [comment]
https://community.libretranslate.com/t/improving-transliteration-in-libretranslate/400

**Comment by argosopentech** [comment]
I have a fix:
https://github.com/LibreTranslate/LibreTranslate/pull/345

@pierotofy

**Comment by jacotec** [comment]
May I just ask the stupid dummy question for what this transliteration is needed when it works without? What will a missing transliteration break?

**Comment by jspablo** [comment]
Same question lol I have tested your fix with a python environment locally and it works

**Comment by PJ-Finlay** [comment]
> May I just ask the stupid dummy question for what this transliteration is needed when it works without? What will a missing transliteration break?

In some cases converting between languages with different phonetic character sets might not work as well without explicit transliteration. For example, if you're translating English to Russian it might translate a name using the English characters instead of the phonetic equivalent in the Russian alphabet. I want to look into [improving transliteration in LibreTranslate](https://community.libretranslate.com/t/improving-transliteration-in-libretranslate/400) but for now the priority is getting LibreTranslate working.

TLDR: Removing transliteration may have slightly worse translations in some cases but it won't error out.

**Comment by PJ-Finlay** [comment]
> Same question lol I have tested your fix with a python environment locally and it works

Thanks @jspablo , I'm going to merge this now but @pierotofy needs to push to PyPI and DockerHub to get this fix widely distributed.

**Comment by jspablo** [comment]
Thanks you all for the fast reponse and solution!

**Comment by pierotofy** [comment]
Just released https://github.com/LibreTranslate/LibreTranslate/releases/tag/v1.3.2 (PyPI will update shortly).

We might be able to setup a mirror for Polyglot data and reintroduce the transliteration logic later on.

**Comment by jacotec** [comment]
Cool! And a new docker hub is also there :-) I've updated the container 10 minutes ago and as soon as the webserver comes online (this always takes ages) I'll test.
@pierotofy @PJ-Finlay @argosopentech  Thanks for the superfast fix!

> We might be able to setup a mirror for Polyglot data and reintroduce the transliteration logic later on.

That sound like a good plan!

**Comment by jacotec** [comment]
Anyone else having issues with the docker hub image? My webinterface does not come up at all. Container running 25 minutes now. Something is wrong with that image ...

**Comment by jacotec** [comment]
@PJ-Finlay Can it be that the now missing transliteration files affect the detection of the language?
I had several Spanish toots on my Mastodon server which, after running the translation, still are unchanged - with the remark "Translated from English by LibreTranslate". Seems it did not correctly detect the language ...

**Comment by delaballe** [comment]
Hi all,

The issue seems fixed

https://polyglot.cs.stonybrook.edu/~polyglot/

is Up & Running Again

Thx a lot for the quick resolution

Cheers

**Comment by PJ-Finlay** [comment]
> @PJ-Finlay Can it be that the now missing transliteration files affect the detection of the language? I had several Spanish toots on my Mastodon server which, after running the translation, still are unchanged - with the remark "Translated from English by LibreTranslate". Seems it did not correctly detect the language ...

I just tested the latest Docker container and language detection is working. It's possible in specific cases it's failing though especially if the text being detected is short.

**Comment by jacotec** [comment]
I've reverted to the 1.3.1 image in Docker Hub and can confirm that it's working (so the @delaballe is right).

But maybe it's better to move the download to your own server or embed them into the image, so not all your installations are depending on the status of a 3rd party server.

**Comment by PJ-Finlay** [comment]
Based on [my tests](https://community.libretranslate.com/t/improving-transliteration-in-libretranslate/400/3) I think using a separate transliteration library may not be necessary. More feedback or tests would also be good to see how transliteration works with vs without Polyglot.
